New model manageiq manageiq provider

yaacov opened the pull request:
**SUMMARY**

ManageIQ is an open source management platform for Hybrid IT.

This change is adding:

    manageiq_provider module, responsible for provider management in ManageIQ

All the modules, including docs, tests and usage examples can be found here

Currently, the only requirement for the modules is manageiq-api-client-python

**ISSUE TYPE**

    New Module Pull Request

**COMPONENT NAME**

manageiq_provider.py (module)

**ANSIBLE VERSION**

ansible 2.3.0.0

yaacov committed:
New Model manageiq manageiq_provider

yaacov commented:
@gtanzillo @abellotti (ManageIQ team), @gundalow @thaumos (Ansible team), @zgalor @cben @simon3z (Containers Management team) - please review.

Reviewers from ManageIQ: Gregg Tanzillo (@gtanzillo) and Alberto Bellotti (@abellotti).
Maintainers: @cben @zgalor @yaacov

zgalor commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,517 @@
+          port: 8443
+          verify_ssl: False
+        metrics:
+          hostname: 'example.com'
```
please add an example with setting `role: prometheus`

yaacov replied:
👍 added

zgalor commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,517 @@
+        Amazon=dict(
+            class_name='ManageIQ::Providers::Amazon::CloudManager',
+            authtype='default',
+            metrics_role='metrics',
```
the default metric role for amazon should be `None`

yaacov replied:
👍 changed

zgalor commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,517 @@
+              'ssl-with-validation-custom-ca',
+              'ssl-without-validation',
+            ]),
+        ca_bundle_path=dict(),
```
`ca_bundle_path` param was replaced by `certificate_authority`. Should be removed

yaacov replied:
👍 right

zgalor commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,517 @@
+        return None
+
+    def zone_id(self, name):
+        """ Search for group id by zone name.
```
/s/group/zone/

yaacov replied:
👍 done

zgalor commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,517 @@
+        # clean nulls, we do not send nulls to the api
+        endpoints = delete_nulls(endpoints)
+
+        # try to create a new user
```
s/user/provider

yaacov replied:
👍 done

zgalor reviewed (changes requested):
Nice refactoring of original module. Few comments.

zgalor reviewed (approved):
LGTM

yaacov commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,528 @@
+        Openshift=dict(
+            class_name='ManageIQ::Providers::Openshift::ContainerManager',
+            authtype='bearer',
+            metrics_role='hawkular',
```
@cben @simon3z WDYT should be the default metrics_role : 'hawkular' or 'prometheus' ?

we currently default to 'hawkular' but thinking of the future we should default to 'prometheus' ?

yaacov replied:
changed to default to 'prometheus' if role is not explicitly set.

yaacov commented:
@gtanzillo @abellotti please review.

yaacov commented:
@cben please review

yaacov commented:
@abellotti @gtanzillo hi, I need your ack / nack before I start nagging the ansible guys :-)

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,523 @@
+
+        # if we do not have a provider_type, use the current provider_type
+        if provider and not provider_type:
+            provider_type = manageiq_provider.class_name_to_type(provider['type'])
```
Nice :+1:
If trying to edit a type not supported by this module, `class_name_to_type()` will return None and will not give clear error message?
Similarly if creating new and omitted type, will it give clear error message?

yaacov replied:
> will it give clear error message?

yes, we will send it to the API, but the error returned is clear and to the point.

yaacov replied:
> If trying to edit a type not supported by this module, class_name_to_type() will return None and will not give clear error message?

provider['type'] can only be `supported_providers().keys()` so it should not be `None`, but even if it is ^^ :-)

cben replied:
`class_name_to_type()` has fallthrough return None if it doesn't find (unless I'm reading it wrong)

yaacov replied:
p.s. added checks for null or bad provider type, checked ... this can happen if we manually tweak the type in the DB, and then send an edit request without explicit type :-)
[EDIT: no this can not happen, but leaving the test, it may help in some other case ... ]

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,523 @@
+        provider_defaults = supported_providers().get(provider_type)
+
+        # build default endpoint
+        endpoint = raw_endpoints.get('default')
```
naming nit: singular `endpoint` is one of `raw_endpoints` from playbook, plural `endpoints` is in "connection_configurations" format.
can you find less confusing names?  perhaps rename `endpoints` list to `configurations`?
(also consider s/build_endpoints/build_configurations/)

yaacov replied:
👍 done, now using the variable "connection_configurations" instead of endpoints.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,523 @@
+        #       it to current state.
+
+        # clean nulls, we do not send nulls to the api
+        resource = delete_nulls(resource)
```
:+1: this works around https://github.com/ManageIQ/manageiq/pull/13400 bug, and possibly helps with older API versions that had less fields (as long playbook doesn't specify the new fields).

Subtle: here you omit top-level None fields, in create only inside endpoints.
Does this allow you to do partial edit?

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,523 @@
+      default[port]:
+        description:
+          - The provider's api port.
+        required: true
```
shouldn't be required, irrelevant for Amazon providers

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,523 @@
+    default: 'default'
+  region:
+    description:
+      - The provider's region name in manageiq.
```
Both ManageIQ and Amazon have "regions" and "zone".  
IINM, it's zone in manageiq DB and AWS region for Amazon.
=> Please clarify descriptions.
=> Consider `manageiq_zone` and/or `provider_region`?

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,523 @@
+          - The CA bundle string with custom certificates. defaults to None.
+'''
+
+EXAMPLES = '''
```
Can you add an Amazon example?

yaacov replied:
👍 done, added an Amazon example ( without port and with provider_region)

yaacov replied:
👍 done :-)

cben reviewed (commented):
Overall neat :smile_cat: 
This is significant simplification compared to what we had in  https://github.com/dkorn/manageiq-ansible-module/blob/master/library/manageiq_provider.py :-)

1. Doesn't wait for ManageIQ to test provider connection — this was complex & fragile, needs better API https://github.com/ManageIQ/manageiq-api/issues/28
2. Doesn't schedule a refresh, IIUC can only be done after (1) above (and possibly would belong in a separate module).
3. Doesn't attempt to guess whether update is needed, this was incorrect due to https://github.com/ManageIQ/manageiq-api/issues/27
  => Always does edit, forced to always assume `changed: true`.  Can't do better without API improvement.

P.S. multi-endpoint "connection_configurations" API is not fully documented, https://github.com/ManageIQ/manageiq-api/issues/26 is best reference I have.

yaacov committed:
update docs, port is not required, region is provider-region

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,523 @@
+
+        # if we do not have a provider_type, use the current provider_type
+        if provider and not provider_type:
+            provider_type = manageiq_provider.class_name_to_type(provider['type'])
```
`supported_providers()` has fallthrough `return None` if it doesn't find (unless I'm reading it wrong)

cben reviewed (approved):
:+1:

@gtanzillo @abellotti Please review.  We especially need your perspective on whether the module interface would work in the future for other provider types.

yaacov commented:
@martinpovolny

gtanzillo reviewed (approved):
👍 LGTM

yaacov commented:
shipit

yaacov commented:
cc @thaumos @mattclay @bcoca

syncrou commented on lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:
```diff
@@ -0,0 +1,540 @@
+      manageiq_connection:
+        url: 'http://127.0.0.1:3000'
+        username: 'admin'
+        password: 'smartvm'
```
@yaacov Can you add and example including the token?

yaacov replied:
happily :-) added at:
https://github.com/ansible/ansible/pull/28273/files#diff-d282374f7f4e8ce8c63becb85a3130b4R226

syncrou replied:
Awesome Thanks!  👍

yaacov committed:
add example of using token

yaacov commented:
shipit

yaacov commented:
@thaumos @mattclay @bcoca please review

yaacov committed:
loop on endpoints instead of creating them one by one

yaacov commented:
shipit

yaacov committed:
add alerts endpoint

yaacov commented:
shipit

yaacov commented:
@thaumos @mattclay @bcoca please review

zgalor commented:
shipit

yaacov commented:
@thaumos @mattclay @bcoca please review

samdoran committed:
Simplify boilerplate and reorganize docs

Tried to make it clearer how the suboptions are laid out.

samdoran commented:
Is there an existing ManageIQ module this one deprecates? If so, do we want to warn on that module to use the new one?

samdoran commented:
I would like to get this merged, but the highly nested options on the `endpoints` field are a bit of a challenge. Accurately documenting them is the first challenge. In the current state, they will not render properly on docs.ansible.com. Second, suboptions more than one level deep tend to indicate overloading a parameter with too much information.

Would it be possible to simplify the module arguments so that so many parameters are not specified in `endpoints`, and move them to top level arguments instead?

I will see about getting this merged in its current state, but I would like to follow up with another PR to simplify the parameters if possible.

samdoran committed:
Flatten out suboptions in order to make tests pass

These will not render properly in the HTML docs. Simplifying how this module accepts options should be addressed in a future PR.

samdoran commented:
Merged to `devel` for inclusion in 2.4. Thank you!

yaacov commented:
> I would like to follow up with another PR to simplify the parameters if possible.

@samdoran thanks for the docs touch-up 👍 

I think we can do:
```
endpoints:
  - role: 'default'
    hostname: 'example.com'
    ....
  - role: 'hawuklar'
    auth_key: 'fdhlfkha lkj'
    ...

```
instead of:
```
endpoints:
  metrics:
    role: 'default'
    hostname: 'example.com'
    ....
  alerts:
    role: 'hawuklar'
    auth_key: 'fdhlfkha lkj'
    ...

```

@cben @enoodle @ilackarms @zgalor @simon3z WDYT ?

simon3z commented:
> @cben @enoodle @ilackarms @zgalor @simon3z WDYT ?

@yaacov I think you have to sync with @gtanzillo because the risk here is that this could be too much container-oriented, what would be the plan for other providers?

Even if the implementation is not there I think we should plan for being able to later accommodate any kind of provider.

yaacov commented:
> @yaacov I think you have to sync with @gtanzillo because the risk here is that this could be too much container-oriented, what would be the plan for other providers?

@simon3z 

All providers have a "default" endpoint, so we are covered here. "default" is the only endpoint that is "required=true" , the other endpoints are per-provider.

We also have a set of pre-defined defaults for each supported provider:
https://github.com/ansible/ansible/pull/28273/files#diff-d282374f7f4e8ce8c63becb85a3130b4R239
When we add a new supported provider, we can also improve the defaults value.

The pros and cons I see in using a list vs. using a dict are:
Pros of list:
1. users can set any set of endpoint they want ( e.g. we do not limit them to one metrics endpoint )
2. it may be more human understable, and will have less indentation levels.

Cons of list:
1. users can set any set of endpoint they want ( e.g. we do not limit them to one metrics endpoint )
2. it's harder code-wise to set default values per `provider-type/option` because we have more possible options and they are not pre-ordered inside a dict.

I like the dict option better, @cben suggested that users may prefer to have one level options ( e.g. "metrics_hostname" instead of `{ "metrics": {"hostname": "example.com"} }` ) because it's more human lagauge like.

cben commented:
Correction: I'm not suggesting `metrics_hostname`, `metrics_port`... that
would be bad.
My proposal is dropping the `endpoints:`, lifting these to top level:

default:
  role: 'default'
  hostname: 'example.com'
  ....
alerts:
  role: 'hawuklar'
  auth_key: 'fdhlfkha lkj'
  ...

perhaps with an optional common suffix/prefix to separate those from other
level-options:


default_endpoint:
  role: 'default'
  hostname: 'example.com'
  ....
alerts_endpoint:
  role: 'hawuklar'
  auth_key: 'fdhlfkha lkj'
  ...




2017-08-30 14:34 GMT+03:00 Yaacov Zamir <notifications@github.com>:

> @yaacov <https://github.com/yaacov> I think you have to sync with
> @gtanzillo <https://github.com/gtanzillo> because the risk here is that
> this could be too much container-oriented, what would be the plan for other
> providers?
>
> @simon3z <https://github.com/simon3z>
>
> All providers have a "default" endpoint, so we are covered here. "default"
> is the only endpoint that is "required=true" , the other endpoints are
> per-provider.
>
> We also have a set of pre-defined defaults for each supported provider:
> https://github.com/ansible/ansible/pull/28273/files#diff-
> d282374f7f4e8ce8c63becb85a3130b4R239
> When we add a new supported provider, we can also improve the defaults
> value.
>
> The pros and cons I see in using a list vs. using a dict are:
> Pros of list:
>
>    1. users can set any set of endpoint they want ( e.g. we do not limit
>    them to one metrics endpoint )
>    2. it may be more human understable, and will have less indentation
>    levels.
>
> Cons of list:
>
>    1. users can set any set of endpoint they want ( e.g. we do not limit
>    them to one metrics endpoint )
>    2. it's harder code-wise to set default values per provider-type/option
>    because we have more possible options and they are not pre-ordered inside a
>    dict.
>
> I like the dict option better, @cben <https://github.com/cben> suggested
> that users may prefer to have one level options ( e.g. "metrics_hostname"
> instead of { "metrics": {"hostname": "example.com"} } ) because it's more
> human lagauge like.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/28273#issuecomment-325963545>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAQtGPqWCSMIwjeP34asJRLzAL5O-Wv5ks5sdUi4gaJpZM4O4v_e>
> .
>

yaacov commented:
@enoodle @ilackarms as the only people I know of that use this module :-)
the options now are:

a. leave this as it is now:
```
endpoints:
   default:
        role: ....
        hostname: ...
   metrics:
        hostname: ....

```
b. use one less level ( as @cben suggested ) with "endpoint" being inferred: 
 ```
default:
      role: ....
      hostname: ...
 metrics:
      hostname: ....

```

this two options sound good to me.

1. do you have an opinion ?
2. do you know of other users who may have an opinion ?

simon3z commented:
@yaacov personally I like "a" (*leave this as it is now*), but maybe I am missing something. I'll leave this to 
 the feedback of the direct users (@ilackarms @joelddiaz @cben)

yaacov commented:
> but maybe I am missing something.

@simon3z , the missing item of knowledge is that ansible **docs** do not know how to **display help about** 3rd level of indentation, the module itself works fine.

I think the module is better with this extra level of indentation, but a user looking at the docs may get confused.

see:
http://docs.ansible.com/ansible/devel/manageiq_provider_module.html

this is why @samdoran  asked that we consider removing one level of indentation.

simon3z commented:
> @simon3z , the missing item of knowledge is that ansible docs do not know how to display help about 3rd level of indentation, the module itself works fine.

@yaacov ah yeah indeed I don't think we should design the API trying to please the docs limitations (unless they were there for a reason, e.g. more than two levels are confusing, etc.)

yaacov commented:
@samdoran hi,
we can fix the extra `suboptions` ( see above discussion ) but we think that "manageiq-wise" it's better the way it is now, how bad is this issue "ansible-wise" ?

samdoran commented:
The docs rendering issue is minor — it's more about the module UI and how it behaves compared to other modules.

This is the first module we have that has _three_ layers of suboptions. Will this be a burden on the user, having to specify a complex nested dictionary? Is there any way the options could be simplified/flattened to top level parameters, possibly defining dependency relationships in order to provide helpful warnings if certain combinations are submitted/omitted?

I'm not saying it _has_ to be this way, but we generally try to streamline the module UI so they are simple to use.

This is a community module, and it may be that the community using this module likes this UI. Ultimately it is up to the community — we just try to give guidance so modules have a consistent look and feel.

yaacov commented:
@samdoran thanks,

@ilackarms @joelddiaz @enoodle @cben , please read @samdoran comment above, please vote for one of this options:

Your feedback is important. 

A.

```
name: ...
type: ...
manageiq_connection:
     url: ...
     token: ...
endpoints:
     default:
         role: ....
         hostname: ...
     metrics:
         hostname: ....
```

B.

```
name: ...
type: ...
manageiq_connection:
     url: ...
     token: ...
default:
     role: ....
     hostname: ...
metrics:
     hostname: ....
```

C.
```
name: ...
type: ...
manageiq_connection:
     url: ...
     token: ...
default_endpoint:
     role: ....
     hostname: ...
metrics_endpoint:
     hostname: ....
```

joelddiaz commented:
The docs do take a moment to get used to (if you're used to reading Ansible module documents). I can see how the manageiq_connection section can be flattened out. Renaming those params to something like miq_username / miq_password / miq_url / etc makes some sense to me. This is more-or-less how the AWS modules work.

The provider section is tougher to flatten out. The parameter naming would end up encoding the parentage of the params. Something like:

The base provider params:
provider_name
provider_type
provider_hostname
provider_ssl
provider_certificate_authority
provider_auth_key

The metrics:
provider_metrics_hostname
provider_metrics_ssl
provider_metrics_certificate_authority

Alerts:
provider_alerts_hostname
provider_alerts_ssl
provider_alerts_certificate_authority

As far as @yaacov 's A/B/C options, I would go with B. But may I suggest one minor change. Instead of default->role, default->hostname, etc, can it be something besides 'default':

provider:    (<--- or 'provider_endpoint')
  role: ...
  hostname: ...

cben commented:
I vote B.

As for flattenning `manageiq_connection: foo: ...` to `miq_foo: ...`, that's what we had in https://github.com/dkorn/manageiq-ansible-module/ but IMHO nesting there has 2 benefits:
- many concepts are relevant both to ansible<->MIQ connection and MIQ<->provider (host, port, user, pw, ssl enforcement); user name in manageiq_user was similarly  ambiguos.  IMHO having them under a separate section is a clearer separation than prefixing `miq_foo: ...` (but YMMV).
- same connection settings are likely to be reused accross manageiq module calls.  having all in one dictionary allows building it once and concisely reusing.

yaacov commented:
@cben Thanks, I also think that's the nicest option :-) , also changed "default" to "provider" as @joelddiaz sugested.

See: #28967
